benchmarks/cpp/src/qsort.cpp

```diff
@@ -0,0 +1,73 @@
+#include <algorithm>
+#include <asio.hpp>
+#include <vector>
+#include <iostream>
+#include <random>
+
+using namespace std::chrono;
+using asio::awaitable;
+using asio::co_spawn;
+using asio::detached;
+
+constexpr int partition(std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    const auto pivot = static_cast<unsigned long>(end - 1);
+    int i = start;
+    for (unsigned long j = start; j < pivot; ++j) {
+        if (arr[j] <= arr[pivot]) {
+            std::iter_swap(arr.begin() + i, arr.begin() + j);
+            ++i;
+        }
+    }
+    std::iter_swap(arr.begin() + i, arr.begin() + pivot);
+    return i;
+}
+
+constexpr void insertionSort(std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    for (unsigned long i = start + 1; i < end; ++i) {
+        for (unsigned long n = i; n > start && arr[n] < arr[n - 1]; --n) {
+            std::iter_swap(arr.begin() + n, arr.begin() + n - 1);
+        }
+    }
+}
+
+void shuffle(std::vector<int>& arr) {
+    std::mt19937 rng(std::random_device{}());
+    std::shuffle(arr.begin(), arr.end(), rng);
+}
+
+awaitable<void> quickSort(asio::io_context& ctx, std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    if (end - start <= 32) {
+        insertionSort(arr, start, end);
+        co_return;
+    }
+
+    const auto pivot = partition(arr, start, end);
+    co_await quickSort(ctx, arr, start, start + pivot);
```

Comment:
This looks like it should be `co_spawn()` for it to run concurrently to the quickSort() after it. Otherwise, it converges into a normal/synchronously quickSort() impl.

Reply:
`asio::co_spawn` is a function that creates a new fiber to execute a function concurrently with the current fiber. It returns an awaitable object that can be awaited to wait for the new fiber to complete.

Reply:
yea, so it should be something like this to allow the other half to run concurrently:
```cpp
auto fiber = co_spawn(ctx/pool, qsort, start, start + pivot)
co_await qsort(start + pivot, end)
co_await fiber;
```

Reply:
> 
> ```c++
> auto fiber = co_spawn(ctx/pool, qsort, start, start + pivot)
> co_await qsort(start + pivot, end)
> co_await fiber;
> ```

It would not work, because `qsort` is void type.

Reply:
I was using pseudo-code, also qsort returns an awaitable instead of void so it should work no?

Reply:
Unfortunately it didn't work here in the tests.
Still researching the overhead involved with `co_spawn`. However, some implicit changes in the asio code after version 1.20 make this difficult.